common/app/common/Logback/LogbackConfig.scala

```diff
@@ -0,0 +1,89 @@
+package common.Logback
+
+import ch.qos.logback.classic.spi.ILoggingEvent
+import ch.qos.logback.classic.{Logger => LogbackLogger, LoggerContext}
+import com.amazonaws.auth.STSAssumeRoleSessionCredentialsProvider
+import com.gu.logback.appender.kinesis.KinesisAppender
+import net.logstash.logback.layout.LogstashLayout
+import org.slf4j.LoggerFactory
+import play.api.LoggerLike
+import conf.Configuration
+
+object LogbackConfig {
+
+  lazy val loggingContext = LoggerFactory.getILoggerFactory.asInstanceOf[LoggerContext]
+
+  lazy val appName: String = Configuration.environment.projectName
+  lazy val customFields = Map(
+    "stack" -> "frontend",
+    "app" -> appName,
+    "stage" -> Configuration.environment.stage.toUpperCase
+  )
+
+  case class KinesisAppenderConfig(stream: String,
+                                   region: String,
+                                   roleArn: String,
+                                   sessionName: String,
+                                   bufferSize: Int)
+
+  def makeCustomFields(customFields: Map[String, String]): String = {
+    "{" + (for((k, v) <- customFields) yield(s""""${k}":"${v}"""")).mkString(",") + "}"
+  }
+
+  def asLogBack(l: LoggerLike): Option[LogbackLogger] = l.logger match {
+    case l: LogbackLogger => Some(l)
+    case _ => None
+  }
+
+  def makeLayout(customFields: String) = {
+    val l = new LogstashLayout()
+    l.setCustomFields(customFields)
+    l
+  }
+
+  def makeKinesisAppender(layout: LogstashLayout, context: LoggerContext, appenderConfig: KinesisAppenderConfig) = {
+    val a = new KinesisAppender[ILoggingEvent]()
+    a.setStreamName(appenderConfig.stream)
+    a.setRegion(appenderConfig.region)
+    a.setCredentialsProvider(new STSAssumeRoleSessionCredentialsProvider(
+      appenderConfig.roleArn, appenderConfig.sessionName
+    ))
+    a.setBufferSize(appenderConfig.bufferSize)
+
+    a.setContext(context)
+    a.setLayout(layout)
+
+    layout.start()
+    a.start()
+    a
+  }
+
+  def initLogger(logger: LoggerLike, config: LogStashConf) = {
+    if (config.enabled) {
+      asLogBack(logger).map { lb =>
+        try {
+          lb.info("Configuring Logback")
```

Comment:
There is actually not one logger but one logger for each class implementing
the Logging Trait (in order to give a logger a more specific and detailed
context I guess). Add to this the default Play.Logger object.
This is not something introduced by this PR, it was already the case before
<https://github.com/guardian/frontend/pull/12475/files/5bd1749af0f7b07df32429f247d5dfc864fa8f5c#diff-367ed9cbe4c8c9e3035658d7fbcaa2eaL6>
.
What this change does is adding an appender to all the loggers when they
are created.

Can you give me more details about the leak you are seeing? Thx



On 16 April 2016 at 23:33, John Duffell <notifications@github.com> wrote:

> In common/app/common/Logback/LogbackConfig.scala
> <https://github.com/guardian/frontend/pull/12475#discussion_r59975179>:
>
> > +    ))
> > +    a.setBufferSize(appenderConfig.bufferSize)
> > +
> > +    a.setContext(context)
> > +    a.setLayout(layout)
> > +
> > +    layout.start()
> > +    a.start()
> > +    a
> > +  }
> > +
> > +  def initLogger(logger: LoggerLike, config: LogStashConf) = {
> > +    if (config.enabled) {
> > +      asLogBack(logger).map { lb =>
> > +        try {
> > +          lb.info("Configuring Logback")
>
> I think this is running more times than it should, every time the Logging
> trait is created it inits the logger which is causing a leak
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/12475/files/5bd1749af0f7b07df32429f247d5dfc864fa8f5c#r59975179>
>



-- 

Thomas Bonnin

Software developer

Guardian News & Media

-----

thomas.bonnin@theguardian.com

-----

Kings Place, 90 York Way,

London N1 9GU

theguardian.com

​-----
​​Download the Guardian app for Android
<https://play.google.com/store/apps/details?id=com.guardian&hl=en_GB> and
iOS​ <https://itunes.apple.com/gb/app/the-guardian/id409128287?mt=8>


-- 

Thomas Bonnin

Software developer

Guardian News & Media

-----

thomas.bonnin@theguardian.com <dominic.kendrick@theguardian.com>

-----

Kings Place, 90 York Way,

London N1 9GU

theguardian.com

​-----
​​Download the Guardian app for Android
<https://play.google.com/store/apps/details?id=com.guardian&hl=en_GB> and
iOS​ <https://itunes.apple.com/gb/app/the-guardian/id409128287?mt=8>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396